src/Avalonia.Visuals/Rendering/DirtyVisuals.cs

```diff
@@ -89,19 +100,69 @@ public bool Remove(IVisual visual)
         /// <returns>A collection of visuals.</returns>
         public IEnumerator<IVisual> GetEnumerator()
         {
-            foreach (var i in _inner)
+            using (DeferChanges())
             {
-                foreach (var j in i.Value)
+                foreach (var i in _inner)
                 {
-                    yield return j;
+                    foreach (var j in i.Value)
+                    {
+                        yield return j;
+                    }
```

Comment:
Would it make more sense if ``GetEnumerator` returned the deferred changes immediately here after it had finished enumerating?